### ISSUE TITLE
[3.0] Roadmap

### ISSUE THREAD

**Issue by barryvdh** [thephpleague/omnipay#274]
Sometimes I see Omnipay V3 being mentioned in the comments, so perhaps best to keep track of it somewhere. So what are the actual changes?

- [ ] Bump PHP version to 5.5?
- [ ] Use PSR-7 instead of Symfony Foundation/Guzzle directly? (#320, https://github.com/thephpleague/omnipay-common/pull/74)
- [ ] Bump Guzzle to v6 or make optional? (Any psr-7 client would do; http://httplug.io/ ?)
- [ ] Way to determine what parameters are needed for a gateway? (#266)
- [ ] Money value object for amount/currency? (https://github.com/thephpleague/omnipay/issues/274#issuecomment-111810211)
- [ ] Namespace change to `League\Omnipay` (https://github.com/thephpleague/omnipay/issues/274#issuecomment-111213052)
- [ ] Split Card in CC info + Customer personal details (https://github.com/thephpleague/omnipay/issues/274#issuecomment-111289516)
- [ ] Storage driver (https://github.com/thephpleague/omnipay/issues/274#issuecomment-111720660)
- [ ] Locale/language standardisation (https://github.com/thephpleague/omnipay-mollie/issues/21)
- [ ] Multiple status values possible (#281)

API cleanup:
- [ ] `isCompleted()` vs `isSuccessful()` (#275)
- [ ] `transactionId` & `transactionReference` confusion?
- To be determined?

Timeline: End 2015?

But as it's probably a big work to update all gateways also, we should probably only update if needed.

Imho, the most interesting part to upgrade would be PSR-7, because it would reduce dependancy with Symfony/Guzzle, as we could use that to: A. construct the request to gateways. B. construct the request back to the developer.

**Comment by cfreear** [comment]
:+1: for 5.5 and psr-7 so that devs can choose their own (psr-7 compatible) http clients.

Further feedback on the 5.5 thing would be beneficial to make sure there isn't an issue for some consumers and their hosts not providing 5.5...

**Comment by kayladnls** [comment]
All that you listed is correct. I want to decouple our dependance on Guzzle and Symfony and PSR-7 would be a good way to do that. 

Also, adding better support for Bank Accounts, which requires breaking up the Customer object from the Card object. 

Better documentation for each of the gateways. I figure as we update gateways, we can write documents as we go so that they each have better documentation.

**Comment by judgej** [comment]
Was there mention of a vendor name change for composer? I kind of like `omnipay` and don't see any advantage of changing that. But I may have just misheard or misunderstood. Maybe it was the namespace root name? Can you put me out of my confused misery? ;-)

**Comment by kayladnls** [comment]
@judgej all League packages have the league namespace. This will be part of V3. 

FQNS would be: `League/Omnipay/Gateway/ClassName`

**Comment by judgej** [comment]
The cart - can we do something with that to allow it to be extended when needed? At the moment it is kind of embedded in middle of omnipay-common and you cannot get to it before it is instantiated and passed to the gateway driver.

Use-case is that for some gateways I would like to be able to extend the parameters that the gateway supports, for a richer set of basket details for gateways that support it. This can assist the service owner with audit trails and reconciling accounts further down the line.

In fact, I wonder if this is a wider issue? Supposing OmniPay used whatever DI container was available for its own internal services. Could that be used as a way to rewire the internals of OmniPay where-ever the need arose, perhaps to put in an extended cart in place of the default one?

**Comment by judgej** [comment]
Thanks - writing up the "Version 3" section for my OmniPay talk for PHPNE on Tuesday, so that's useful to know.

**Comment by kayladnls** [comment]
Is it being recorded? I'd love to see!

**Comment by judgej** [comment]
It isn't normally, but I'll ask if anyone would like to do the camera work. I'll put all the slides and notes up on github anyway.

**Comment by judgej** [comment]
@cfreear I personally thing `transactionId` and `transactionReference` are fine. I do realise there is at least one gateway that uses these exact same names for opposite purposes, but I suspect that will happen somewhere along the line whatever names are chosen. That is also mainly an issue for the gateway driver developer, and should not concern the users of the drivers too much.

I think this is a documentation issue. The difference between those two fields should be explained in a big orange callout somewhere near the top of the documentation. That's my 2p.

**Comment by judgej** [comment]
Sorry not trying to hijack this thread - last post for today ;-) This is something for V3 that has been mentioned a few times:

* Splitting the Card class in separate Card and Customer/Payee classes. There are times when you need either sets of data without the other, and the validation on the half not needed gets in the way of that.

**Comment by maxnet** [comment]
Talking about ``transactionReference``.
Think that when receiving a ``RedirectResponse`` getting and saving the ``transactionReference`` to pass it to ``CompletePurchaseRequest`` later on should be made a hard requirement.


With some modules it is optional right now, and they fallback to insecure behavior using GET parameters the customer can modify (and point to a different previous transaction that did succeed).
E.g. Mollie's completePurchaseRequest:

```php
        $data['id'] = $this->getTransactionReference();

        if (!isset($data['id'])) {
            $data['id'] = $this->httpRequest->request->get('id');
        }
```

**Comment by judgej** [comment]
@maxnet Ouch. There are too many examples of gateway plugins that do that type of thing, often in the notify handler just returning "yeah, whatever" while not actually checking the security hashes or matching them against the stored transaction.

I think OmniPay needs a standard storage interface that perhaps drivers could use. That storage would be indexed by the transactionId (plus some other values for multiple captures against payments), the gateway status and transaction state (i.e. overall status of where it is at), plus custom fields for the application use (e.g. I store all message data in all directions, which is invaluable for tracing issues later). I started a package that could do this, but kind of got stuck and needed to move on.

The idea would be similar to PSR-7 - "I want to send and receive HTTP messages, so give me the tools to do so". In this case it would be the tools to store, retrieve and amend transactions in non-volatile storage.

**Comment by barryvdh** [comment]
Yeah I agree, part of the problem is that is just isn't really clear how a gateway is supposed to behave. Partially also because of differences in the gateways.

**Comment by judgej** [comment]
Should we be taking these suggestions off to separate issues, and give them a "3.0 Ideas" label?

I think this storage one is a great idea. Having the right interface available for drivers that do want to use it, could work really work well. Not all drivers would need it, as they don't involve access across multiple sessions, or involve the user's browser at all (i.e. direct APIs) but those that do, really should use the storage to keep checking things have not been changed.

**Comment by barryvdh** [comment]
Or extend the interface, eg. to verify a complete Purchase, you would need to provide the original reference as parameter.

**Comment by maxnet** [comment]
>The idea would be similar to PSR-7 - "I want to send and receive HTTP messages, so give me the tools to 
>do so". In this case it would be the tools to store, retrieve and amend transactions in non-volatile storage. Is 
>that along the kind of lines you were thinking?

Yes.

Having a database, might also be useful to support gateways that require you to poll transaction status periodically.
(with ING iDeal Advanced they send you an e-mail telling you off, if you did create a transaction, but your software never polled its status afterwards. Can happen if customer does not click link to get redirected back to site.).

**Comment by maxnet** [comment]
>Or extend the interface, eg. to verify a complete Purchase, you would need to provide the original reference as parameter.

That's currently already possible (see Mollie), but providing the reference would need to be made a hard requirement.

**Comment by judgej** [comment]
> Or extend the interface, eg. to verify a complete Purchase, you would need to provide the original reference as parameter.

Some already do this, such as SagePay's CompleteAuthorize::send(). However, you can completely skip that stage and return a SUCCESS regardless of whether the TransactionReference (containing the hashes) match what has been sent in the notification. Someone not wanting to store the transaction may be tempted to do that. They may also just do `exit("OK\r\nRedirectUrl=http://example.com/paid-successfully");` and skip OmniPay totally, but we can't help them ;-) What we want to do, is make it easier to do it the right way to encourage it.

**Comment by barryvdh** [comment]
Making it a hard requirement would make it obvious it needs to be stored somewhere (and should be noted that session doesn't suffice) but perhaps we should investigate to see if there are gateways wo don't support it. Perhaps all just support it and it wouldn't be a problem 

Storage driver could also work, but not sure if that makes it easy enough when people store custom information or account for gateway differences.

**Comment by maxnet** [comment]
>Storage driver could also work, but not sure if that makes it easy enough when people store custom 
information or account for gateway differences.

Just stick all the custom information in a TEXT field as json?

**Comment by judgej** [comment]
@maxnet yes, that's what I do. Usually I have about five columns exposed as indexed/searchable columns - transactionId, amount, overallStatus, gatewayStatus, errorMessage and a few others, and then everything else goes into a TEXT column. ORMs like eloquent can silently handle all the JSON encoding and decoding so you don't even know its there. An adapter between a flexible enough interface and storage can handle all the necessary mapping. The overallStatus tracks what happens to the transaction as far as the application is concerned - did it register, is it waiting for a notification, has it been aborted, did it get rejected, is it complete with a successul payment etc. The driver statuses do not generally reflect this in a consistent way. For example, SagePay returns "OK" when a transaction has been successfully registered, and it returns "OK" when a payment has been successfully authorised. That can be kind of confusing, because that same status means two different things, while an overallStatus provides some context to that.

It is something I've wanted to see anyway for a while, just to take away the laborious task of having to decide what to store each time on different gateways, and how to store it on different frameworks - there is *nothing* out there to just pick up and run with.

**Comment by judgej** [comment]
I think the storage needs for all the drivers can be pretty much standardised. There are only a few pieces of information necessary for the functionality (hashes to check, authorisation status to pass out from a notification handler), and the rest is just custom fields. I've seen gateway plugins that tell the application whether the payment was authorized *solely* by sending the user to a `success` or `fail` URL, avoiding having to store this detail. Makes the hair on the back of neck stand up. *shiver*

**Comment by judgej** [comment]
Any thoughts about using a value object for the monetary values? That object could encapsulate all the necessary validation, formatting, would have a defined currency etc. For backwards compatibility, the old methods could remain, but would just feed into the money value object. Whether it would be built into OmniPay, or provided by a third-party package, or offered as an interface plus an adapter for a recommended third party package(s) with the option to drop others in, I'm not sure. Personally, I've had too many hairy dealings with the `amount` values being passed in to remain comfortable with it stying as a string.

**Comment by kayladnls** [comment]
@judgej If we're going to use anything for Money, I'd prefer to use this: https://github.com/mathiasverraes/money
Thoughts?

**Comment by lord2800** [comment]
I'm personally not a fan of Omnipay making storage decisions (re: transaction ids/refs/messages) for me--any reason not to depend on an interface and then have a few well-known implementations?

**Comment by judgej** [comment]
That seems to be the one that all money roads lead to. The next release is a complete rewrite, and looks much lighter: https://github.com/mathiasverraes/money/tree/nextrelease

**Comment by ivank** [comment]
I can't wait for the guzzle / psr7 in this update - would be great. Guys have you considered https://github.com/sebastianbergmann/money for money representation? (That's the guys who made phpunit) It looks a bit more "stable" and thought out.

**Comment by judgej** [comment]
@ivank This money package has some nice practical features for converting between different formats and units. I do like it.

I think, so long as Omnipay provides its own interface for what it expects of a Money object, then simple adapters can be written for any money package. This is similar to the PSR-7 approach - rather than land on Guzzle for communications and lock the package into Guzzle, just accepting a PSR-7 interface is enough for the end developer to use *any* HTTP communications package they like,

**Comment by barryvdh** [comment]
So, a concrete problem arises:
 
- Omnipay2 requires Guzzle3
- Guzzle3 requires Symfony 2 (event dispatcher), and is EOL
 - Symfony 3 is already released.

So we can't use Omnipay with Symfony 3..\

Upgrading to Guzzle 6 will require that all gateways are converted to use PSR-7 messages, which is gonna take some time..

**Comment by judgej** [comment]
I assume once PSR-7 messages are being used, then any PSR-7 client can be used to do the comms, perhaps with an appropriate connector? I've been trying out that approach on a [Particle (IoT) API](https://github.com/academe/phpParticle/tree/master/src) to see how practical it is and what I can learn from it, and it seems to be possible. That's not to say it will not be a lot of work. And that's assuming I understand just what the aims of moving to PSR-7 really is about.

**Comment by barryvdh** [comment]
That is the idea, we don't rely on Guzzle message interfaces, but PSR-7 Message interfaces.
We also don't really need the http foundation responses, but can return PSR-7 responses.

Once Guzzle6 updates to Guzzle7, we should be able to migrate without replacing all gateways, if they respect the interface. (Or to any other PSR-7 compatible http client)

**Comment by judgej** [comment]
What I've found is that there is some handy stuff that Guzzle does with the messages in the background, such as adding headers automatically. That is low-level stuff that needs to be handled by the application using the PSR-7 messages. I don't think pushing that to all the gateways is the way to go, as it spreads too much HTTP details around the drivers. A central OmniPay-PSR7 implementation is probably the way to go, offering the messages (maybe as connectors to a strict PSR-7 implementation) but adding a layer on top (is that a decorator?) to implement the stuff that PSR-7 does not do, but which we know in practice all the gateways will need.

I'm probably not explaining myself very well, but hopefully it makes sense :-)

**Comment by barryvdh** [comment]
You mean for creating a Request/Response object?

Perhaps so, we have to create those objects somewhere anyways, and if we use `guzzle/psr7`, we're still tied to Guzzle. Not sure if that should be a decorator or a factory or  something else.

```php
// Factory
$request = Omnipay\RequestFactory::get('http://my-domain.com', ['a' => 123]);
$request = Omnipay\RequestFactory::post('http://my-domain.com', $headers, $body);

$response = Omnipay\ResponsFactory::string($body);
$response = Omnipay\ResponseFactory::redirect($url);
```

Should we open a new ticket for this and evaluate some options?

**Comment by makasim** [comment]
In payum I introduced a [http client interface](https://github.com/Payum/Payum/blob/master/src/Payum/Core/HttpClientInterface.php) and [guzzle as default implementation](https://github.com/Payum/Payum/blob/master/src/Payum/Core/Bridge/Guzzle/HttpClient.php).

There is [tiny factory](https://github.com/Payum/Payum/blob/master/src/Payum/Core/Bridge/Guzzle/HttpClientFactory.php) to build an http client faster

**Comment by judgej** [comment]
Yes, that's right. `guzzle/psr7` `Request` and `Response` classes seem to be a pretty "pure" implementation of PSR-7, with little that goes beyond the PSR. I expect they could be used pretty much directory, so an OmniPay `Request` object could probably extend.a Guzzle PSR-7 `Request` object and be used pretty much directly.

Also not sure the best way - factory, connector, decorator or a mix of them all. It's knowing where to draw the lines that is unclear without some examples. For example, headers will be needed for some GET requests, because the headers can carry authentication tokens. Now, whether the gateways should be dealing with headers directly, or passing "Bearer" tokens to OmniPay to put the headers in, I really don't know.

 It has lots of helpers for creating streams, so you wouldn't use those without a connector of some sort if you don't want to be tied in.

**Comment by judgej** [comment]
I like the general idea of keeping the generation of the PSR-7 messages separate from the sending and receiving of PSR-7 responses though. It feels right that building data, and then using it over HTTP, are two very different things that should be kept apart.

**Comment by judgej** [comment]
Bridge, Connector - I feel like I have to go back to school to relearn the subtleties between them and what they do...

**Comment by delatbabel** [comment]
As long as we can pass arbitrary headers in, and in addition cater for gateways (First Data Webservices) that require server certificate + client key + password for establishment of the SSL connection. For the current gateway I had to drop out of guzzle and do it directly in cURL.

**Comment by barryvdh** [comment]
There is also a pretty popular implementation here: https://github.com/zendframework/zend-diactoros

Guzzle psr-7 [Request](https://github.com/guzzle/psr7/blob/master/src/Request.php): 
`__construct($method, $uri, array $headers = [], $body = null,$protocolVersion = '1.1' )`
Diactoros [Request](https://github.com/zendframework/zend-diactoros/blob/master/src/Request.php): 
`__construct($uri = null, $method = null, $body = 'php://temp', array $headers = [])`

Which could be just a function in the AbstractRequest, right? Or possibly abstract it to a request factory, but that's probably overkill. As long as people rely on the PSR interface, not the actual implementation, it should be easy to swap out.

```php
public function createRequest($method, $uri, $headers = [], $body = null) {
    return new \GuzzleHttp\Psr7\Request($method, $uri, $headers, $body);
    return new \Zend\Diactoros\Request($uri, $method, $body, $headers);
    return new RequestFactory::createRequest($method, $uri, $headers, $body);
}```

**Comment by judgej** [comment]
Yes. I personally think PSR-7 is too low level for the gateways - it puts too much burden on the developer to understand HTTP to a low level (e.g. PSR-7 knows absolutely nothing about how to format the body of a POST request to send data - we don't want 30 implementations of that done by 30 different developers). How I would see it, is that OmniPay would provide an implementation, or a series of implementations that implement an interface, which does more-or-less what Guzzle or Diactoros does. Behind the scenes they would pass as much as they can to Guzzle or whatever, and they would offer some additional helpers such as addBearerToken, addPostParameters etc.

So I think OmniPay should rely on the PSR-7 interface only, when dealing with the messages created by Guzzle and friends, but provide more than that to the gateway drivers. A gateway driver *could* deal with PSR-7 messages if it wants, but PSR-7+OmniPay features would be *much* nicer to deal with.

That is just mu conclusion having played with PSR-7 messages over the last few weeks - it is *so* low level, that you really need more if you are not going to get bogged down in HTTP intricacies.

**Comment by judgej** [comment]
Taking [this as an example](https://github.com/academe/phpParticle/blob/master/src/ParticleApi.php#L644) that method generates PSR-7 messages to send to an Arduino device. These are plain GET, PUT, DELETE messages, some involving uploading files, and also involving several different types of authentication (OAuth and device tokens). That's all standard stuff - nothing special - but look at all those headers and stuff that needs to be handled in there. That is the kind of thing that dealing with pure PSR-7 messages will force you to do. That is why this stuff, IMO, should be offered by OmniPay as decorated/extended PSR-7 messages so all that stuff can be wrapped up out of the way.

**Comment by barryvdh** [comment]
We'll have to check for common headers/patterns, but to be honest, it shouldn't be too hard for gateway developers to configure a couple of headers.. And gateways might have slight differences in implementations anyways..

Another thing, should we aim for an easy upgrade path, or don't care?
Should we rename the package to `league/omnipay`, dump the omnipay-common package and just put the Common stuff in the League\Omnipay package? I don't really see the use for the global 'require all gateways' package..

**Comment by barryvdh** [comment]
Let's continue the PSR/Request stuff in #320

**Comment by barryvdh** [comment]
Shall we also simplify the getting of the Gateway, since we now have the `::class` operator?

Instead of `Omnipay::create('Stripe')` and converting/matching the gateway name, just:

```php
use Omnipay\Stripe\StripeGateway;
Omnipay::create(StripeGateway::class);
```
And also remove the stuff about finding/listing all gateways from composer etc.

**Comment by aimeos** [comment]
No, that would prevent creating gateways by configuration!

**Comment by barryvdh** [comment]
Why?
`Omnipay::create('Omnipay\Stripe\StripeGateway');` would also work, so you can store that instead of `Stripe`

**Comment by aimeos** [comment]
This makes it more difficult for users to get it right. We have experienced a lot of users failing on such small things which only increases the amount of support requests. Please always try to keep things as simple as possible.

**Comment by barryvdh** [comment]
Using the class name = the simplest. Your IDE can autocomplete it and you don't have to guess.
Right now it uses some automatic transformation to move to the Omnipay namespace.
They should also just use any namespace they want.

**Comment by aimeos** [comment]
Think about the users, not only the developers. In Aimeos shop owners can configure the Omnipay gateways using some settings like the gateways name (currently e.g. "Stripe"). They are no developers and know nothing about namespaces and maybe haven't seen an IDE in their entire life.

For the developer point of view (mine) it doesn't make things easier either. The factory Omnipay:create() should abstract from the concrete instantiation and shouldn't bother me with implementation details. The concrete class name is such an implementation detail which can change, the string "Stripe" should not.

Implementation of the factory would be easier for sure if the concrete class name is given but than you don't need the factory either ...

**Comment by barryvdh** [comment]
Omnipay is meant for developers, not for end-users.
If you want simple names, just map them in your back-end to the real names.
Stripe is just as much imlementation detail, it can only map to `Omnipay\Stripe\Gateway`, so can't change either.
The point of the factory is (in my opininion) to inject the proper dependancies.

**Comment by aimeos** [comment]
You are only looking at your use case and that's totally OK. But there are other valid use cases that can be made more or less difficult. Your proposed change would make it much more difficult for us and for many others I've seen here in the past who have asked how Omnipay drivers can be instantiated by configuration.

A mapping is static and the worst option only suitable if you are out of any other option. Requiring a static mapping for the factory would be the exact opposite of what a factory is made for: A factory is a central instance that creates objects without the need to know the implementation details of how to create the objects AND the dependencies of the created objects.

From Wikipedia: "... factory methods deal with the problem of creating objects without having to specify the exact class of the object that will be created" (https://en.wikipedia.org/wiki/Factory_method_pattern)

**Comment by maxnet** [comment]
>Omnipay is meant for developers, not for end-users.

Disagree with that.
If #266 ever gets implemented Omnipay should be universal enough that the user can switch to a different payment module without hiring a developer.


>And also remove the stuff about finding/listing all gateways from composer etc.

Why?
Would rather see it fixed to properly auto-detect installed gateways.

You may be able to use Composer's API for that.
As in require gateways to add the gateway name to the extra field of their own package's ``composer.json`` instead of in omnipay-common.
Get a list of installed packages with ``$composer->getRepositoryManager()->getLocalRepository()->getPackages()``, call ``getExtra()`` on the objects.

**Comment by judgej** [comment]
At the moment, the gateway name [maps directly to a class](https://github.com/thephpleague/omnipay-common/blob/5fa50e04275d354bceb0870c0bf1cb6421b3a6f7/src/Omnipay/Common/Helper.php#L106) - the factory is just a handy way to replace underscores and prepend the class with `\Omnipay` and append with `Gateway`. In other words, given an "official" OmniPay gateway name, its entry-point class can be derived from that name directly.

However, you can also already pass in your own custom class, and that works without any modifications.

It looks to me like everyone is served here, so should be happy :-) well, kind of.

The real issue here, if I'm understanding this correctly, is getting a list of which gateways are installed and available for use. @maxnet if you have any examples of a package of framework that is able to read the composer information, that would be a great help. I've never seen this before, and have tried investigating whether it can be done (about a year ago) and just could not get into the composer API - at the time you needed to install the composer source as a package, and that alone was an awkward thing to do - maybe it's all easier now).

The OmniPay example app reads the composer.json file to get a list, which is really not good (it assumes it knows where composer.json is, and it's a fixed list of what the installed version of omnipay-common happens to have been told about, plus it only sees "official" gateways.

------

So, I would suggest leaving the factory to operate as it does. It currently *does* accept classes, so anyone can use those if they want. It also accepts official gateway names, and maps them onto a class as a handy feature, but no more than just a handy renaming.

Ideally there would be a way to as composer which gateways are installed - official or not - and that IMO is worth perusing.

**Comment by barryvdh** [comment]
Well, you could create some registration file (`register.php`), which runs `OmnipayRegistry::register('Stripe', 'Omnipay\Stripe\Gateway')`, so that you have a Registry with gateways. You can autoload that file so it's always run (that is how Magento 2 handles modules I think), but it isn't really pretty imho.

And yes, we can load them by classname already, but only with a `\` prefix, which doesn't work for the `::class` resolution.

**Comment by judgej** [comment]
Maybe a quick fix to look for a `\` *anywhere* in the name and do a `class_exists()` on it would be a step in the right direction. That would support `::class` and the old names and absolute class paths too.

**Comment by barryvdh** [comment]
Yeah as long as you don't have a class with the same name, which isn't that case _most_ of the times..

**Comment by judgej** [comment]
*most* times? It always worries me when a client says, "Unequivocally no, that will never happen, ever. Except sometimes, but not often enough to worry about."

**Comment by barryvdh** [comment]
Yeah I also meant it wasn't a good idea, because we can't be sure people don't have a local Stripe\Paypal etc class.

**Comment by judgej** [comment]
It would also need to implement an expected interface and/or have an expected public method before we take it as one of ours.

**Comment by lord2800** [comment]
Something to keep in mind is `composer.json` doesn't have to be deployed with the application--it's only needed at build-time. Perhaps a "best of both worlds" solution here is to have a command line utility that can generate a "registry" of sorts? That way you can just run a command to generate a list of installed gateways.

**Comment by judgej** [comment]
`composer.json` is only a starting point when installing an application anyway, and does not give you the list of what *actually* got installed, and all their dependencies.

**Comment by delatbabel** [comment]
I agree that there needs to be a simple mapping from a gateway string to a class name, and the current mapping, e.g. PayPal_Rest -> Omnipay\PayPay\RestGateway works well. There are many other applications where the gateway short name and configuration parameters are stored in a database configuration allowing admins (not developers) to switch to a new gateway by fiddling with the config through an admin UI and not. Requiring a full class name would be too much of a BC break.

**Comment by delatbabel** [comment]
There are methods to pull out the registered class names other than looking at composer.json. Checking the array returned by requiring vendor/composer/autoload_classmap.php is the most common method I've seen.

**Comment by Jellyfrog** [comment]
"`transactionId` vs `transactionReference`" still isn't very clear, I have a gateway that returns both `id` and `paymentReference`, but it uses `id` for what Omnipay calls "`transactionReference`" because its what you use to check status/refund and so on...
Maybe it should be prefixed with gateway/merchant? For example; `gatewayTransactionId`, `gatewayTransactionReference`, `merchantTransactionId`, this would make it more clear what comes from where...

**Comment by aimeos** [comment]
Why not name the local ID simply "orderid"? This is what it is and everybody knows the term.

**Comment by judgej** [comment]
Documentation should make it clear - there are two field names, each with a distinctive purpose, and that should not be hard to understand. @aimeos Introducing orderid is just going to confuse things more. I have one order, and make two attempts at paying it - do both attempts now get the same internal ID? Or is *that* order id not actually *this* order id? Confused already.

Many gateways also have an order id field that is used to hold an optional order id, for systems that have orders, with ids. Don't mix them up.

**Comment by judgej** [comment]
@Jellyfrog the point of OmniPay is that it provides the translation between what the remote gateway calls fields and what the local application should call those fields when talking to OmniPay. The rule is set that OmniPay calls your local application payment ID the `transactionId` and the ID used internally to the gateway the `transactionReference`. That's the convention that is laid out, and the remote gateways - dozens of them - can call those fields whatever they like. Sometimes they will have similar names, sometimes reversed names, sometimes completely different names, and sometimes not even specify a name for `transactionId` forcing the gateway to make up its own name and put it into a custom field (e.g. Authorize.net).

tl;dr: Yes, these two fields will almost never have the same names as your remote gateway calls them, but *that is the point of OmniPay* - to translate.

**Comment by kayladnls** [comment]
This conversation has gotten massive and is covering a lot of things. I like the aggregation but I think we need to break a lot of these points out into their own issues. 

I created the V3.0 label to keep track of these things. Let's open issues for each topic at hand to keep the conversation more focused and easy to follow.

**Comment by barryvdh** [comment]
Thanks. I've create an issue for the Money stuff: https://github.com/thephpleague/omnipay/issues/322

**Comment by barryvdh** [comment]
I've also bumped to PHP 5.6, because that has extended support, while 5.5 is almost over.

**Comment by aimeos** [comment]
If you think of enterprise distributions (e.g. RHEL), they have a lifetime of 5-7 years. They support old versions of PHP longer than the official team.

I would suggest to only carefully raise the bar for PHP versions and only if you really need the new features, e.g. to 5.4 if traits are required. If there are no such requirements, please leave it to PHP 5.3. There's no value in requiring new versions only because they are officially not supported any more.

**Comment by barryvdh** [comment]
It was 5.5 already for v3. And 5.5 is already not more supported, only for security issues.
